server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -172,7 +176,57 @@ boolean isJoinPending() {
         return pendingOutgoingJoins.iterator().hasNext();
     }
 
-    void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
+    private static class FailedJoinAttempt {
+        private final DiscoveryNode destination;
+        private final JoinRequest joinRequest;
+        private final TransportException exception;
+        private final long timestamp;
+
+        FailedJoinAttempt(DiscoveryNode destination, JoinRequest joinRequest, TransportException exception) {
+            this.destination = destination;
+            this.joinRequest = joinRequest;
+            this.exception = exception;
+            this.timestamp = System.nanoTime();
+        }
+
+        void maybeLogNow() {
+            if (isSuspiciousTransportException(exception)) {
+                logger.info(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            } else {
+                logger.debug(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            }
+        }
+
+        boolean isSuspiciousTransportException(TransportException e) {
+            if (e instanceof RemoteTransportException) {
+                Throwable cause = e.getCause();
+                if (cause != null &&
+                        cause instanceof CoordinationStateRejectedException  ||
+                        cause instanceof FailedToCommitClusterStateException ||
+                        cause instanceof NotMasterException) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        void logWarnWithTimestamp() {
+            logger.info(() -> new ParameterizedMessage("last failed join attempt was {} ms ago, failed to join {} with {}",
+                            TimeValue.timeValueMillis(TimeValue.nsecToMSec(System.nanoTime() - timestamp)),
+                            destination,
+                            joinRequest),
+                    exception);
+        }
+    }
+
+
+    void logLastFailedJoinAttempt() {
+        if (lastFailedJoinAttempt != null) {
+            lastFailedJoinAttempt.logWarnWithTimestamp();
```

Comment:
I think this can throw a NPE because we read the volatile field twice.

Reply:
Good catch! https://github.com/elastic/elasticsearch/pull/39756/commits/0b19f03a7160c3a4a0fbfc1351ca028519c44a9b